File: 2024/day_22/solution.py
from typing import List

def generate_gift_sets(gifts: List[str]) -> List[List[str]]:
    result = []

    # Helper function to generate combinations of size k
    def generate_combinations(start: int, current_combo: List[str], size: int):
        if len(current_combo) == size:
            result.append(current_combo[:])
            return

        for i in range(start, len(gifts)):
            current_combo.append(gifts[i])
            generate_combinations(i + 1, current_combo, size)
            current_combo.pop()

    # Generate combinations of all sizes from 1 to n
    for size in range(1, len(gifts) + 1):
        generate_combinations(0, [], size)

    return result

File: 2024/day_22/solution.js
function generateGiftSets(gifts) {
  const result = [];

  // Helper function to generate combinations of size k
  function generateCombinations(start, currentCombo, size) {
    if (currentCombo.length === size) {
      result.push([...currentCombo]);
      return;
    }

    for (let i = start; i < gifts.length; i++) {
      currentCombo.push(gifts[i]);
      generateCombinations(i + 1, currentCombo, size);
      currentCombo.pop();
    }
  }

  // Generate combinations of all sizes from 1 to n
  for (let size = 1; size <= gifts.length; size++) {
    generateCombinations(0, [], size);
  }

  return result;
}

console.log(generateGiftSets(["car", "doll", "puzzle"]));
// [
//   ['car'],
//   ['doll'],
//   ['puzzle'],
//   ['car', 'doll'],
//   ['car', 'puzzle'],
//   ['doll', 'puzzle'],
//   ['car', 'doll', 'puzzle']
// ]

console.log(generateGiftSets(["ball"]));
// [
//   ['ball']
// ]

console.log(generateGiftSets(["game", "pc"]));
// [
//   ['game'],
//   ['pc'],
//   ['game', 'pc']
// ]
